### PR TITLE
Revert/add 43862 dynamic template names descriptions

### PR DESCRIPTION
**Pull request by cbravobernal**
## What?

#43862 will finally be landed in 6.1. This PR adds the review comments done in the PR by @dream-encode

### CODE REVIEW

**Comment by dream-encode** [comment]
@c4rl0sbr4v0 I reverted the revert of https://core.trac.wordpress.org/changeset/54333 in https://core.trac.wordpress.org/changeset/54370.

The JS side of this is in the GB packages, and will need another build sync for 6.1 Beta 3 tomorrow.

**Comment by Mamaduka** [comment]
This morning I was looking through #43862 and forgot to add the comment 🤦

I've some concerns about how this change might affect the `get_block_templates` function performance, especially on setups where many templates require dynamic names/descriptions.

Based on my simple tests, I see increased query numbers when running the `gutenberg_get_block_templates`.

## Testing instructions
* 5 custom templates for authors.
* 5 custom templates for posts.
* Run example test below before #43862 got merged (1645fb2292ce2064b6a74f09be1069b794106cca) and after (e4087ae814aeffb29c375c2ee9ac5028f5a6b96a).

## Simple method to test query numbers
```php
add_action( 'init', function() {
	global $wpdb;

	// Reset query count.
	$wpdb->num_queries = 0;

	gutenberg_get_block_templates();

	error_log( "Queries: $wpdb->num_queries" );
} );
```

**Comment by cbravobernal** [comment]
I'm removing the backport label to prevent the automation to cherry pick this today.

**Comment by ockham** [comment]
cc/ @jorgefilipecosta @priethor Can I ask y'all to have a look into [those potential performance issues](https://github.com/WordPress/gutenberg/pull/44645#issuecomment-1265765498) that @Mamaduka found?

**Comment by jorgefilipecosta** [comment]
> This morning I was looking through #43862 and forgot to add the comment 🤦
> 
> I've some concerns about how this change might affect the `get_block_templates` function performance, especially on setups where many templates require dynamic names/descriptions.
> 
> Based on my simple tests, I see increased query numbers when running the `gutenberg_get_block_templates`.
> 
> ## Testing instructions
> * 5 custom templates for authors.
> * 5 custom templates for posts.
> * Run example test below before [Update: Make template names and descriptions dynamic. #43862](https://github.com/WordPress/gutenberg/pull/43862) got merged ([1645fb2](https://github.com/WordPress/gutenberg/commit/1645fb2292ce2064b6a74f09be1069b794106cca)) and after ([e4087ae](https://github.com/WordPress/gutenberg/commit/e4087ae814aeffb29c375c2ee9ac5028f5a6b96a)).
> 
> ## Simple method to test query numbers
> ```
> add_action( 'init', function() {
> 	global $wpdb;
> 
> 	// Reset query count.
> 	$wpdb->num_queries = 0;
> 
> 	gutenberg_get_block_templates();
> 
> 	error_log( "Queries: $wpdb->num_queries" );
> } );
> ```

Hi @Mamaduka, @ockham,

I think it is expected that if we want the template titles to reflect the real titles of a post, author or taxonomy, we will need queries to retrieve these titles from the database and in the end there is no way around that. To reflect the more updated information and react to changes on the base entity we need to query the entities. So having more queries in that function is something expected.

What we can do is thing like adding a parameter to gutenberg_get_block_templates where we can disable title computation if the caller of the function does not requires it (e.g: just wants to render a template), or see if get_user_by is more performant than get_user as @Mamaduka suggested. I guess we can do these optimizations on the next release if we see there is an impact we can minimize.

By default on most websites (where no entity specific templates exist) there is no change on the number of queries at all. When the user creates entity specific templates we have some additional queries to compute the names on the next release according to the feedback we receive we can see if we can find some improvements.

**Comment by ockham** [comment]
Thank you for addressing those concerns @jorgefilipecosta!

**Comment by ockham** [comment]
I'll add back the "Backport to WP Beta/RC" label. Since I've already published npms for Beta 3 (which is due in about 1.5 hrs), I'd rather not rush it in. Instead, we'll include it in RC1 next week.

(Note that the [PHP code has already been backported](https://github.com/WordPress/gutenberg/pull/44645#issuecomment-1265730423) -- which [shouldn't be a problem](https://wordpress.slack.com/archives/C03LZ88NX6G/p1664887575708039?thread_ts=1664450838.412229&cid=C03LZ88NX6G).)

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 1]
```diff
@@ -248,6 +248,136 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
...
+ *
+ * @param string            $post_type Post type e.g.: page, post, product.
+ * @param string            $slug      Slug of the post e.g.: a-story-about-shoes.
+ * @param WP_Block_Template $template  Template to mutate adding the description and title computed.
+ * @return boolean Returns true if the referenced post was found and false otherwise.
+ */
+function _gutenberg_build_title_and_description_for_single_post_type_block_template( $post_type, $slug, WP_Block_Template $template ) {
+	$post_type_object = get_post_type_object( $post_type );
+
+	$posts = get_posts(
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/block-template-utils.php:267]: Let's use WP_Query here.

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 2]
```diff
@@ -248,6 +248,136 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
...
+		array(
+			'name'      => $slug,
+			'post_type' => $post_type,
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/block-template-utils.php:270]: Limit query to 1.

Comment (spacedmonkey) [reply lib/compat/wordpress-6.1/block-template-utils.php:270]: Also do not prime term and meta caches, if not used.

File: lib/compat/wordpress-6.1/block-template-utils.php [hunk 3]
```diff
@@ -248,6 +248,136 @@ function gutenberg_get_block_template( $id, $template_type = 'wp_template' ) {
...
+ * @return boolean True if a term referenced was found and false otherwise.
+ */
+function _gutenberg_build_title_and_description_for_taxonomy_block_template( $taxonomy, $slug, WP_Block_Template $template ) {
+	$taxonomy_object = get_taxonomy( $taxonomy );
+
+	$terms = get_terms(
+		array(
+			'taxonomy'   => $taxonomy,
+			'hide_empty' => false,
+			'slug'       => $slug,
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/block-template-utils.php:335]: Limit to 1. 
Do not prime term meta caches.

**Comment by ockham** [comment]
Per discussion [here](https://wordpress.slack.com/archives/C03LZ88NX6G/p1664900718762719?thread_ts=1664792761.838559&cid=C03LZ88NX6G), I'll merge this and will re-publish the npm package.

**Comment by ockham** [comment]
@spacedmonkey's suggested PHP changes are being addressed in https://github.com/WordPress/wordpress-develop/pull/3404.

**Comment by ockham** [comment]
This was included in WP 6.1 Beta 3, see https://github.com/WordPress/wordpress-develop/pull/3405.

**Pull request merged**